identity/backend/src/main/java/io/camunda/identity/rolemanagement/service/RoleMembershipService.java

```diff
@@ -0,0 +1,127 @@
+    final CamundaUserDetails camundaUserDetails =
+        camundaUserDetailsManager.loadUserByUsername(
+            userService.findUserById(userId).getUsername());
+
+    return camundaUserDetails.getRoles();
+  }
+
+  public void assignRoleToUser(final String roleName, final long userId) {
+    if (!roleRepository.existsById(roleName)) {
+      throw new RuntimeException("role.notFound");
+    }
+
+    final CamundaUserDetails camundaUserDetails =
+        camundaUserDetailsManager.loadUserByUsername(
+            userService.findUserById(userId).getUsername());
+
+    final List<String> roles = new ArrayList<>();
+    roles.addAll(camundaUserDetails.getRoles());
+    roles.add(roleName);
+
+    saveUserDetailsWithRoles(camundaUserDetails, roles);
+  }
+
+  public void unassignRoleFromUser(final String roleName, final long userId) {
+    if (!roleRepository.existsById(roleName)) {
+      throw new RuntimeException("role.notFound");
+    }
+
+    final CamundaUserDetails camundaUserDetails =
+        camundaUserDetailsManager.loadUserByUsername(
+            userService.findUserById(userId).getUsername());
+
+    final List<String> roles = new ArrayList<>();
+    roles.addAll(
+        camundaUserDetails.getRoles().stream().filter(rn -> !rn.equals(roleName)).toList());
+
+    saveUserDetailsWithRoles(camundaUserDetails, roles);
+  }
+
+  public List<String> getRolesOfGroupByGroupId(final Long groupId) {
+    final CamundaGroup camundaGroup = groupService.findGroupById(groupId);
+
+    return camundaUserDetailsManager.findGroupAuthorities(camundaGroup.name()).stream()
+        .map(GrantedAuthority::getAuthority)
+        .map(authorityName -> authorityName.replace("ROLE_", ""))
+        .toList();
+  }
+
+  public void assignRoleToGroup(final String roleName, final long groupId) {
+    if (!roleRepository.existsById(roleName)) {
+      throw new RuntimeException("role.notFound");
+    }
+    final CamundaGroup camundaGroup = groupService.findGroupById(groupId);
+
+    camundaUserDetailsManager.addGroupAuthority(
+        camundaGroup.name(), new SimpleGrantedAuthority("ROLE_" + roleName));
+  }
+
+  public void unassignRoleFromGroup(final String roleName, final long groupId) {
+    if (!roleRepository.existsById(roleName)) {
+      throw new RuntimeException("role.notFound");
+    }
+    final CamundaGroup camundaGroup = groupService.findGroupById(groupId);
+
+    camundaUserDetailsManager.removeGroupAuthority(
+        camundaGroup.name(), new SimpleGrantedAuthority("ROLE_" + roleName));
+  }
+
+  private void saveUserDetailsWithRoles(
+      final CamundaUserDetails camundaUserDetails, final List<String> roles) {
+    final UserDetails userDetails =
+        User.withUsername(camundaUserDetails.getUsername())
+            .password(camundaUserDetails.getPassword())
+            .passwordEncoder(Function.identity())
+            .roles(roles.toArray(roles.toArray(new String[0])))
```

Comment:
❓ I'm a little confused here, we do `roles.toArray()` inside another `roles.toArray()`?

Reply:
Valid reason to be confused ^^ ... not sure how that worked, tests ran through in both cases 🙈 ... fixed it though.